lib/utils/buildHelper.js

```diff
@@ -0,0 +1,214 @@
+const log = require("@ui5/logger").getLogger("cli:utils:buildHelper");
+
+/**
+ * Creates an object containing the flattened project dependency tree. Each dependency is defined as an object key while
+ * its value is an array of all of its transitive dependencies.
+ *
+ * @param {object} tree Project tree as generated by the [@ui5/project.normalizer]{@link module:@ui5/project.normalizer}
+ * @returns {object<string, string[]>} An object with dependency names as key and each with an array of its transitive
+ *   dependencies as value
+ */
+function getFlattenedDependencyTree(tree) {
+	const dependencyInfo = {};
+
+	function _getTransitiveDependencies(project, dependencies) {
+		project.dependencies.forEach((dep) => {
+			if (!dependencies.includes(dep.metadata.name)) {
+				dependencies.push(dep.metadata.name);
+				_getTransitiveDependencies(dep, dependencies);
+			}
+		});
+		return dependencies;
+	}
+	function _processDependencies(project) {
+		project.dependencies.forEach((dep) => {
+			if (!dependencyInfo[dep.metadata.name]) {
+				dependencyInfo[dep.metadata.name] = _getTransitiveDependencies(dep, []);
+				_processDependencies(dep);
+			}
+		});
+	}
+
+	_processDependencies(tree);
+	return dependencyInfo;
+}
+
+/**
+ * Creates dependency lists for 'includedDependencies' and 'excludedDependencies'. Regular expressions are directly
+ * applied to a list of all project dependencies so that they don't need to be evaluated in later processing steps.
+ * Generally, includes are handled with a higher priority than excludes. Additionally, operations for processing
+ * transitive dependencies are handled with a lower priority than explicitly mentioned dependencies. The default
+ * dependencies set in the build settings are appended in the end.
+ *
+ * The priority of the various dependency lists is applied in the following order:
+ * <ol>
+ *   <li>includeDependency, includeDependencyRegExp</li>
+ *   <li>excludeDependency, excludeDependencyRegExp</li>
+ *   <li>includeDependencyTree</li>
+ *   <li>excludeDependencyTree</li>
+ *   <li>defaultIncludeDependency, defaultIncludeDependencyRegExp, defaultIncludeDependencyTree</li>
+ * </ol>
+ *
+ * @param {object} parameters Parameters
+ * @param {object} parameters.tree Project tree as generated by the
+ *   [@ui5/project.normalizer]{@link module:@ui5/project.normalizer}
+ * @param {Array.<string>} parameters.includeDependency The dependencies to be considered in 'includedDependencies'; the
+ *   "*" character can be used as wildcard for all dependencies and is an alias for the CLI option "--all"
+ * @param {Array.<string>} parameters.includeDependencyRegExp Strings which are interpreted as regular expressions
+ *   to describe the selection of dependencies to be considered in 'includedDependencies'
+ * @param {Array.<string>} parameters.includeDependencyTree The dependencies to be considered in 'includedDependencies';
```

Comment:
the dot is long obsolete in JSDoc. Use `Array<string>` instead of `Array.string>`. Some linters even prefer string[] over Array<string> when the component type is simple (no union or so).

Reply:
Done